File: backend/internal/models/support.go
package models

import "time"

// Direção da mensagem.
const (
	DirectionIn  = "in"
	DirectionOut = "out"
)

// Ciclo de vida do ticket. "Novo" × "em atendimento" é derivado de
// assigned_user_id (null = ninguém assumiu ainda).
const (
	TicketStatusOpen     = "open"     // novo ou em atendimento
	TicketStatusPending  = "pending"  // aguardando o cliente responder
	TicketStatusResolved = "resolved" // resolvido (reabre se o cliente voltar)
	TicketStatusClosed   = "closed"   // encerrado (nova mensagem cria outro ticket)
)

// ValidTicketStatus indica se o valor é um status conhecido do ciclo de vida.
func ValidTicketStatus(s string) bool {
	switch s {
	case TicketStatusOpen, TicketStatusPending, TicketStatusResolved, TicketStatusClosed:
		return true
	}
	return false
}

// SupportContact é o cliente final atendido (dono do número de WhatsApp).
type SupportContact struct {
	ID        string
	AccountID string
	Phone     string
	Name      *string
	Groups    []ContactGroupRef // grupos de marketing (preenchido na listagem)
	Tags      []SupportTag      // etiquetas do contato (filtro de campanha)
	CreatedAt time.Time
	UpdatedAt time.Time
}

// ContactGroupRef é a referência leve de um grupo (chip no contato).
type ContactGroupRef struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

// ContactGroup é um grupo de contatos (lista de marketing).
type ContactGroup struct {
	ID      string `json:"id"`
	Name    string `json:"name"`
	Members int    `json:"members"` // quantos contatos estão no grupo
}

// ContactGroupRequest cria/renomeia um grupo.
type ContactGroupRequest struct {
	Name string `json:"name" binding:"required,min=1"`
}

// SetContactGroupsRequest substitui os grupos de um contato.
type SetContactGroupsRequest struct {
	GroupIDs []string `json:"group_ids"`
}

// SetContactTagsRequest substitui as etiquetas de um contato.
type SetContactTagsRequest struct {
	TagIDs []string `json:"tag_ids"`
}

// SupportTicket é uma conversa (com protocolo) de um contato.
type SupportTicket struct {
	ID             string
	AccountID      string
	ContactID      string
	Protocol       string
	Status         string
	AssignedUserID *string
	SectorID       *string
	LastMessageAt  time.Time
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// SupportMessage é uma mensagem da conversa.
type SupportMessage struct {
	ID         string
	AccountID  string
	TicketID   string
	Direction  string
	Type       string
	Content    *string
	MediaURL   *string
	MimeType   *string
	FileName   *string
	Status     string
	ExternalID *string
	SenderID   *string
	SenderName *string // nome do atendente (join; exibido nas notas internas)
	Internal   bool    // nota interna: só a equipe vê, nunca vai à Meta
	// Cobrança da Meta: qual modelo foi enviado e em que categoria
	// (marketing/utility/authentication). Vazio nas mensagens comuns.
	TemplateName     *string
	TemplateCategory *string
	CreatedAt  time.Time
}

// --- Respostas do inbox ---

// SupportTicketListItem é uma linha da lista de conversas (com dados do contato).
type SupportTicketListItem struct {
	ID               string    `json:"id"`
	Protocol         string    `json:"protocol"`
	Status           string    `json:"status"`
	ContactName      *string   `json:"contact_name,omitempty"`
	ContactPhone     string    `json:"contact_phone"`
	LastMessageAt    time.Time `json:"last_message_at"`
	AIPaused         bool      `json:"ai_paused"`    // Atendente IA pausado nesta conversa
	UnreadCount      int       `json:"unread_count"` // mensagens recebidas ainda não lidas
	AssignedUserID   *string   `json:"assigned_user_id,omitempty"`
	AssignedUserName *string   `json:"assigned_user_name,omitempty"`
	SectorID         *string   `json:"sector_id,omitempty"`
	SectorName       *string   `json:"sector_name,omitempty"`
	Tags             []SupportTag `json:"tags,omitempty"`
}

// --- Setores (filas de atendimento) ---

// SupportSector é uma fila de atendimento da conta (comercial, suporte…).
type SupportSector struct {
	ID        string    `json:"id"`
	AccountID string    `json:"-"`
	Name      string    `json:"name"`
	Members   []string  `json:"members"` // ids dos atendentes do setor
	CreatedAt time.Time `json:"created_at"`
}

// SectorRequest cria/edita um setor.
type SectorRequest struct {
	Name    string   `json:"name" binding:"required,min=1"`
	Members []string `json:"members"`
}

// --- Eventos do ticket (auditoria + métricas) ---

// Tipos de evento do ticket.
const (
	TicketEventAssigned      = "assigned"       // alguém assumiu/foi atribuído
	TicketEventTransferred   = "transferred"    // mudou de atendente e/ou setor
	TicketEventStatusChanged = "status_changed" // mudou de status
	TicketEventNote          = "note"           // nota interna
	TicketEventReopened      = "reopened"       // cliente respondeu ticket resolvido
)

// SupportTicketEvent é uma entrada do histórico da conversa.
type SupportTicketEvent struct {
	ID             string    `json:"id"`
	Kind           string    `json:"kind"`
	ActorUserID    *string   `json:"actor_user_id,omitempty"`
	ActorName      *string   `json:"actor_name,omitempty"`
	FromUserID     *string   `json:"from_user_id,omitempty"`
	FromUserName   *string   `json:"from_user_name,omitempty"`
	ToUserID       *string   `json:"to_user_id,omitempty"`
	ToUserName     *string   `json:"to_user_name,omitempty"`
	FromSectorID   *string   `json:"from_sector_id,omitempty"`
	FromSectorName *string   `json:"from_sector_name,omitempty"`
	ToSectorID     *string   `json:"to_sector_id,omitempty"`
	ToSectorName   *string   `json:"to_sector_name,omitempty"`
	FromStatus     *string   `json:"from_status,omitempty"`
	ToStatus       *string   `json:"to_status,omitempty"`
	Note           *string   `json:"note,omitempty"`
	CreatedAt      time.Time `json:"created_at"`
}

// --- Fase 2: respostas rápidas, etiquetas e presença ---

// QuickReply é um atalho de texto da empresa (ex.: /boleto → mensagem pronta).
type QuickReply struct {
	ID        string    `json:"id"`
	Shortcut  string    `json:"shortcut"`
	Content   string    `json:"content"`
	CreatedAt time.Time `json:"created_at"`
}

// QuickReplyRequest cria/edita uma resposta rápida.
type QuickReplyRequest struct {
	Shortcut string `json:"shortcut" binding:"required,min=1"`
	Content  string `json:"content" binding:"required,min=1"`
}

// SupportTag é uma etiqueta da empresa (aplicável às conversas).
type SupportTag struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Color string `json:"color"`
	// Uso da etiqueta (preenchido na listagem de gestão).
	Contacts int `json:"contacts,omitempty"`
	Tickets  int `json:"tickets,omitempty"`
}

// TagRequest cria uma etiqueta.
type TagRequest struct {
	Name  string `json:"name" binding:"required,min=1"`
	Color string `json:"color"`
}

// SetTicketTagsRequest substitui as etiquetas de uma conversa.
type SetTicketTagsRequest struct {
	TagIDs []string `json:"tag_ids"`
}

// Presenças possíveis do atendente.
const (
	PresenceAvailable = "available"
	PresenceAway      = "away"
)

// TransferTicketRequest transfere a conversa (atendente e/ou setor, com nota).
type TransferTicketRequest struct {
	UserID   *string `json:"user_id"`   // novo atendente (null mantém/tira)
	SectorID *string `json:"sector_id"` // novo setor
	Note     string  `json:"note"`
}

// SetTicketStatusRequest muda o status da conversa.
type SetTicketStatusRequest struct {
	Status string `json:"status" binding:"required"`
	Note   string `json:"note"`
}

// SupportMessageResponse é a representação pública de uma mensagem.
type SupportMessageResponse struct {
	ID         string    `json:"id"`
	Direction  string    `json:"direction"`
	Type       string    `json:"type"`
	Content    *string   `json:"content,omitempty"`
	MediaURL   *string   `json:"media_url,omitempty"`
	MimeType   *string   `json:"mime_type,omitempty"`
	FileName   *string   `json:"file_name,omitempty"`
	Status     string    `json:"status"`
	Internal   bool      `json:"internal,omitempty"`
	SenderName *string   `json:"sender_name,omitempty"`
	CreatedAt  time.Time `json:"created_at"`
}

// ToResponse converte a mensagem para a resposta pública.
func (m *SupportMessage) ToResponse() SupportMessageResponse {
	return SupportMessageResponse{
		ID:         m.ID,
		Direction:  m.Direction,
		Type:       m.Type,
		Content:    m.Content,
		MediaURL:   m.MediaURL,
		MimeType:   m.MimeType,
		FileName:   m.FileName,
		Status:     m.Status,
		Internal:   m.Internal,
		SenderName: m.SenderName,
		CreatedAt:  m.CreatedAt,
	}
}

// SendMessageRequest é o corpo para o atendente responder no ticket.
type SendMessageRequest struct {
	Content string `json:"content" binding:"required,min=1"`
}

// --- Contatos (cadastro dos clientes finais) ---

// ContactResponse é a representação pública de um contato.
type ContactResponse struct {
	ID        string            `json:"id"`
	Phone     string            `json:"phone"`
	Name      *string           `json:"name,omitempty"`
	Groups    []ContactGroupRef `json:"groups,omitempty"`
	Tags      []SupportTag      `json:"tags,omitempty"`
	CreatedAt time.Time         `json:"created_at"`
}

// ToResponse converte o contato para a resposta pública.
func (c *SupportContact) ToResponse() ContactResponse {
	return ContactResponse{ID: c.ID, Phone: c.Phone, Name: c.Name, Groups: c.Groups, Tags: c.Tags, CreatedAt: c.CreatedAt}
}

// CreateContactRequest cadastra um contato manualmente.
type CreateContactRequest struct {
	Name  string `json:"name" binding:"required,min=1"`
	Phone string `json:"phone" binding:"required,min=8"`
}

// UpdateContactRequest edita nome/telefone de um contato.
type UpdateContactRequest struct {
	Name  *string `json:"name" binding:"omitempty,min=1"`
	Phone *string `json:"phone" binding:"omitempty,min=8"`
}

File: app/lib/models/support.dart
import '../core/phone.dart' show formatPhone;
/// Uma conversa na lista do inbox.
class TicketListItem {
  TicketListItem({
    required this.id,
    required this.protocol,
    required this.status,
    required this.contactPhone,
    this.contactName,
    required this.lastMessageAt,
    this.aiPaused = false,
    this.unreadCount = 0,
    this.assignedUserId,
    this.assignedUserName,
    this.sectorId,
    this.sectorName,
    this.tags = const [],
  });

  final String id;
  final String protocol;
  String status; // open | pending | resolved | closed (mutável p/ ações no header)
  final String contactPhone;
  final String? contactName;
  final DateTime lastMessageAt;
  bool aiPaused; // Atendente IA pausado nesta conversa (mutável p/ toggle otimista)
  int unreadCount; // mensagens recebidas não lidas (mutável p/ zerar ao abrir)
  String? assignedUserId; // atendente responsável (null = ninguém assumiu)
  String? assignedUserName;
  String? sectorId; // setor/fila (null = sem setor)
  String? sectorName;
  List<TicketTag> tags;

  /// Copia os campos mutáveis vindos de uma resposta da API (claim/transfer/status).
  void applyFrom(TicketListItem o) {
    status = o.status;
    assignedUserId = o.assignedUserId;
    assignedUserName = o.assignedUserName;
    sectorId = o.sectorId;
    sectorName = o.sectorName;
    tags = o.tags;
  }

  /// Rótulo humano do status ("Novo" quando aberta e sem dono).
  String get statusLabel => switch (status) {
        'open' => assignedUserId == null ? 'Novo' : 'Em atendimento',
        'pending' => 'Aguardando cliente',
        'resolved' => 'Resolvido',
        'closed' => 'Fechado',
        _ => status,
      };

  String get displayName => (contactName != null && contactName!.isNotEmpty) ? contactName! : contactPhone;

  /// Telefone formatado — mesmo padrão dos contatos: +55 (21) 99333-9504.
  String get prettyPhone => formatPhone(contactPhone);

  String get initials {
    final s = displayName.trim();
    if (s.isEmpty) return '?';
    final parts = s.split(RegExp(r'\s+'));
    if (parts.length == 1) return parts.first.substring(0, 1).toUpperCase();
    return (parts.first[0] + parts.last[0]).toUpperCase();
  }

  factory TicketListItem.fromJson(Map<String, dynamic> j) => TicketListItem(
        id: j['id'],
        protocol: j['protocol'] ?? '',
        status: j['status'] ?? 'open',
        contactPhone: j['contact_phone'] ?? '',
        contactName: j['contact_name'],
        lastMessageAt: DateTime.tryParse(j['last_message_at'] ?? '')?.toLocal() ?? DateTime.now(),
        aiPaused: j['ai_paused'] == true,
        unreadCount: (j['unread_count'] as num?)?.toInt() ?? 0,
        assignedUserId: j['assigned_user_id'],
        assignedUserName: j['assigned_user_name'],
        sectorId: j['sector_id'],
        sectorName: j['sector_name'],
        tags: ((j['tags'] as List?) ?? const [])
            .map((e) => TicketTag.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// Etiqueta da empresa (aplicável às conversas).
class TicketTag {
  TicketTag({required this.id, required this.name, required this.color});

  final String id;
  final String name;
  final String color; // hex, ex.: #0E9384

  factory TicketTag.fromJson(Map<String, dynamic> j) =>
      TicketTag(id: j['id'], name: j['name'] ?? '', color: j['color'] ?? '#0E9384');
}

/// Resposta rápida (atalho de texto): /boleto → mensagem pronta.
class QuickReply {
  QuickReply({required this.id, required this.shortcut, required this.content});

  final String id;
  final String shortcut;
  final String content;

  factory QuickReply.fromJson(Map<String, dynamic> j) =>
      QuickReply(id: j['id'], shortcut: j['shortcut'] ?? '', content: j['content'] ?? '');
}

/// Um setor (fila de atendimento) da empresa.
class Sector {
  Sector({required this.id, required this.name, this.members = const [], this.adDefault = false});

  final String id;
  final String name;
  final List<String> members; // ids dos atendentes do setor

  /// Setor que recebe os leads vindos de anúncio (Click-to-WhatsApp). Um por empresa.
  final bool adDefault;

  factory Sector.fromJson(Map<String, dynamic> j) => Sector(
        id: j['id'],
        name: j['name'] ?? '',
        members: ((j['members'] as List?) ?? const []).cast<String>(),
        adDefault: j['ad_default'] ?? false,
      );
}

/// Uma entrada do histórico (timeline) da conversa.
class TicketEvent {
  TicketEvent({
    required this.kind,
    this.actorName,
    this.fromUserName,
    this.toUserName,
    this.fromSectorName,
    this.toSectorName,
    this.fromStatus,
    this.toStatus,
    this.note,
    required this.createdAt,
  });

  final String kind; // assigned | transferred | status_changed | note | reopened
  final String? actorName;
  final String? fromUserName;
  final String? toUserName;
  final String? fromSectorName;
  final String? toSectorName;
  final String? fromStatus;
  final String? toStatus;
  final String? note;
  final DateTime createdAt;

  static String _statusPt(String? s) => switch (s) {
        'open' => 'Em atendimento',
        'pending' => 'Aguardando cliente',
        'resolved' => 'Resolvido',
        'closed' => 'Fechado',
        _ => s ?? '',
      };

  /// Frase pronta para exibir na timeline.
  String get describe {
    final by = actorName != null ? ' por $actorName' : '';
    switch (kind) {
      case 'assigned':
        return '${toUserName ?? 'Atendente'} assumiu a conversa';
      case 'transferred':
        final parts = <String>[];
        if (toUserName != null) parts.add('para ${toUserName!}');
        if (toSectorName != null) parts.add('para o setor ${toSectorName!}');
        if (toUserName == null && toSectorName != null) {
          return 'Devolvida à fila do setor ${toSectorName!}$by';
        }
        return 'Transferida ${parts.join(' ')}$by';
      case 'status_changed':
        return 'Status: ${_statusPt(fromStatus)} → ${_statusPt(toStatus)}$by';
      case 'reopened':
        return 'Reaberta — o cliente respondeu depois de resolvida';
      case 'note':
        return 'Nota$by';
      default:
        return kind;
    }
  }

  factory TicketEvent.fromJson(Map<String, dynamic> j) => TicketEvent(
        kind: j['kind'] ?? '',
        actorName: j['actor_name'],
        fromUserName: j['from_user_name'],
        toUserName: j['to_user_name'],
        fromSectorName: j['from_sector_name'],
        toSectorName: j['to_sector_name'],
        fromStatus: j['from_status'],
        toStatus: j['to_status'],
        note: j['note'],
        createdAt: DateTime.tryParse(j['created_at'] ?? '')?.toLocal() ?? DateTime.now(),
      );
}

/// Uma mensagem da conversa.
class Message {
  Message({
    required this.id,
    required this.direction,
    required this.type,
    this.content,
    this.mediaUrl,
    this.mimeType,
    this.fileName,
    required this.status,
    required this.createdAt,
    this.internal = false,
    this.senderName,
  });

  final String id;
  final String direction; // in (cliente) | out (atendente)
  final String type; // text | image | document | audio | video | template
  final String? content;
  final String? mediaUrl;
  final String? mimeType;
  final String? fileName;
  final String status;
  final DateTime createdAt;
  final bool internal; // nota interna: só a equipe vê (nunca vai ao cliente)
  final String? senderName; // atendente que enviou (exibido nas notas)

  bool get isOutbound => direction == 'out';
  bool get isImage => type == 'image';
  bool get isAudio => type == 'audio';
  bool get hasMedia => mediaUrl != null && mediaUrl!.isNotEmpty;

  factory Message.fromJson(Map<String, dynamic> j) => Message(
        id: j['id'],
        direction: j['direction'] ?? 'in',
        type: j['type'] ?? 'text',
        content: j['content'],
        mediaUrl: j['media_url'],
        mimeType: j['mime_type'],
        fileName: j['file_name'],
        status: j['status'] ?? '',
        createdAt: DateTime.tryParse(j['created_at'] ?? '')?.toLocal() ?? DateTime.now(),
        internal: j['internal'] == true,
        senderName: j['sender_name'],
      );
}
